### PR TITLE
[WIP] add estimate API in LightningModule

### LINKED ISSUE #2524: [Discussion] What about inference package for PL?
## 🚀 Feature

Inference packages for various models. as @justusschock mentioned in https://github.com/PyTorchLightning/pytorch-lightning/pull/2522#pullrequestreview-442800377, this feature can be called in `trainer.test` to improve test performance. This feature could be implemented by each researcher, but maybe we can discuss "Frequently used functions" and support for the user who is not familiar with these techniques.

IMO, this package if we focus on `p(y|x)` - single model, we can be include

1.0 Inference Improvement 

  - Quite a lot of papers use tricks in the test phase for increase performance
    - Segmentation: multi-scale and image flipping ["HRNET-OCR"](https://arxiv.org/pdf/2005.10821v1.pdf) 

1.1 Support stochastic model ( multiple inference )

  - [Monte Carlo Dropout](https://arxiv.org/pdf/1506.02142.pdf)

2.0 Output Calibration
- Actually this isn't strictly related to metric improvement but can be used for before ensemble - [Temperature Scaling](https://arxiv.org/pdf/1706.04599.pdf)

### Motivation

While Data Competition, tired of writing inference improvement code for each project. If pl includes these, researchers can save their time.

### Pitch

Kaggler/Researcher can improve their test result easily.

### PR DESCRIPTION
**Pull request by davinnovation**
## What does this PR do?

This PR suggests an estimate API for pytorch-lightning.

Expected Behavior with this api
1. Estimate Interactively
2. pre-built techniques for increase estimate result
3. Labeling new data

### Requirements of this API
- consider pytorch/serve

WIP ...

Fixes #2524
Related also https://github.com/PyTorchLightning/pytorch-lightning/issues/1853

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/inference/infer.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.inference.functional import infer as func_infer
+
+
+class InferenceABC(torch.nn.Module, ABC):
+    def __init__(self, network: LightningModule, infer_mode='eval'):
+        super().__init__()
+        self.network = network
+        self.infer_mode = infer_mode
+        self.infer = lambda x: func_infer.infer(self.network, x, self.infer_mode)
```
Comment (justusschock) [inline pytorch_lightning/inference/infer.py:16]: please don't use lambdas as they cannot be pickled

File: pytorch_lightning/inference/infer.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        optim.step(opt)
+
+        self.T = float(T)
+
+
+class IterativeInference(InferenceABC):
+    def __init__(self, network, N:int = 1, infer_mode = 'train', return_mode = 'avg', **kwargs):
+        super().__init__(network, infer_mode, **kwargs)
+        RETURN_MODE = {
+            'avg': lambda x: torch.mean(torch.stack(x), 0),
```
Comment (justusschock) [inline pytorch_lightning/inference/infer.py:66]: again lambdas

File: pytorch_lightning/inference/infer.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            'list': lambda x: x
+        }
+ 
+        self.N = 1
```
Comment (justusschock) [inline pytorch_lightning/inference/infer.py]: variable names should be lowercase

**Comment by davinnovation** [comment]
Thanks for kind comments and review @justusschock . I wrote discussion issue - https://github.com/PyTorchLightning/pytorch-lightning/issues/2524, and will stop progress before the discussion.

File: pytorch_lightning/inference/infer.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+from abc import ABC, abstractmethod
+
+import torch
+
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.core.decorators import auto_move_data
+from pytorch_lightning.inference.functional import infer as func_infer
+
+
+class InferenceABC(torch.nn.Module, ABC):
```
Comment (Borda) [inline pytorch_lightning/inference/infer.py:10]:
```suggestion
class InferenceBase(torch.nn.Module, ABC):
```

File: pytorch_lightning/inference/infer.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+class Inference(InferenceABC):
+    def __init__(self, network, **kwargs):
+        super().__init__(network, **kwargs)
+
+    def forward(self, x):
+        return self.infer(x)
+
+
+class TempScalingInference(InferenceABC):
+    def __init__(self, network, T: float=1.5, **kwargs):
```
Comment (Borda) [inline pytorch_lightning/inference/infer.py:32]: pls follow pep8

File: tests/inference/test_infer.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        self.embed = nn.Embedding(10, 20)   # expects dtype long as input
+        self.reduce = nn.Linear(20, 1)      # dtype: float
+        self.example_input_array = torch.tensor([[0, 2, 1], [3, 5, 3]])  # dtype: long
+
+    def forward(self, x):
+        return self.reduce(self.embed(x))
+
+
+@pytest.fixture
+def random():
```
Comment (Borda) [inline tests/inference/test_infer.py:28]: there already seeding function `seed_everything`

**Comment by davinnovation** [comment]
IMO, inference API can be included in `LightningModule.utils.inference` or `LightningModule.inference`. It is a Draft so we can freely discuss :) Current code is just concept.

1. Inference Interactively without effort ( in Jupyter Hub )

Current PL, inference got lines of code
```python
# Do Trainer.fit(model) ~ Trainer.test() ...
x = torch.Tensor(Want to Predict) 
model.forward = auto_move_data(model.forward) # auto_move_data decorater isn't default :( 

with torch.no_grad(): # without no_grad, sometime CUDA explodes
  model.unfreeze() # stochastic model
  pred0 =model(x)

model.freeze()
pred1 = model(x)

with torch.no_grad(): # without no_grad, sometime CUDA explodes
  model.unfreeze() # stochastic model
  pred2 = torch.mean(torch.stack([model(x) for _ in range(10)]) , 0)
# .. plot or do something
``` 

with Inference API
```python
# Do Trainer.fit(model) ~ Trainer.test() ...
x = torch.Tensor(Want to Predict)
pred0 = model.Inference(x, 'train')
pred1 = model.Inference(x, 'eval')
pred2 = model.IterativeInference(x, 10, 'train', return_mode='avg')
# ... plot or do something
```

2. Inference API can be used in `LightningModule.test_step` and `LightningModule.val_step`. it's not part of `Trainer.test`, it's a utility function for `LightningModule.{}_step`
```python
def validation_step(self, batch, batch_idx):
  # validation_step automatically run with 'eval()' mode
  x, y = batch
  # with torch.no_grad():
  #   model.unfreeze()
  #   pred = self(batch)
  pred = self.Inference(x, 'train')
```

After writing this comment, supporting additional functions `IterativeInferce`, `TemperatureScaling`, `SegmentationAggregate` might be overkill for PL. But inferencing with `unfreeze` option in `validation`, `test` can be useful.

**Comment by jeremyjordan** [comment]
I'm glad to see you've been putting some thought into this @davinnovation!

A lot of thought has gone into developing this library to enable simple training of models based on the Lightning model convention. However, I think that we want to give more thought as to what Lightning looks like when deployed. For example, defining datasets can be unnecessary once you have a trained model artifact and just want to run inference. In some cases, your dataset may fail because the model runtime environment doesn't have access to the same compute resources as the model training environment. (Fortunately the addition of the `setup()` method prevents a Lightning model from failing via this mode in production). 

This PR feels like a step in the right direction. The main difference between what I see proposed here and the existing `test()` functionality is that when running the model over a test dataset when want to collect the predictions and calculate (and log) evaluation metrics. When running inference in production, we simply want to return the prediction for each individual request.

We should also ensure that this works seamlessly with [pytorch/serve](https://github.com/pytorch/serve/).

**Comment by AlexSchuy** [comment]
Any progress on this? It'd be great to have inference integrated with PL.

**Comment by davinnovation** [comment]
I'll resume this project
